### PR TITLE
make db:drop behaviour consistent when a DB does not exists

### PR DESCRIPTION
**Pull request by GustavoCaso**
### Summary

Resolves issue: #32914.

Provides consistent behaviour when using `rake db:drop`

With PostgreSQL and sqlite3 adapters running db:drop with a non-created DB acts as if the DB is there:
Dropped database '<env>'. As opposed to in case of mysql2 adapter: Database '<env>' does not exist.

The solution uses the new `database_exists?` method for the adapters to check if the `DB` before trying to get a connection t it.

If the `DB` does not exist it will return make avoiding the call to `ActiveRecord::Base.connection.migration_context.current_environment` which for some adapters it will create the DB and for others don't.

So finally when trying the `drop` the `DB` we will have the same consistent message across all adapters.

https://github.com/rails/rails/blob/master/activerecord/lib/active_record/tasks/database_tasks.rb#L194-L195

```ruby
rescue ActiveRecord::NoDatabaseError
  $stderr.puts "Database '#{db_config.database}' does not exist"
```

### CODE REVIEW

**Comment by Edouard-chin** [comment]
I don't think there is an issue in Rails. The main problem as you mentioned 

> With PostgreSQL and sqlite3 adapters running db:drop with a non-created DB acts as if the DB is there

This is because in sqlite3 when we initialize a database and don't specify a mode. The mode used by default by sqlite 3 is `SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE` which basically is "The database is opened for reading and writing, and is created if it does not already exist".

If a user don't want his database to be created in case it doesn't exist, he should use `SQLITE_OPEN_READWRITE`. Sqlite3 will use this mode if `readwrite: true` is passed https://github.com/sparklemotion/sqlite3-ruby/blob/b6a9d731c2004d96e6b458c661d156b6a94a4f58/lib/sqlite3/database.rb#L77

so TL;DR to fix that issue people just need to update their database.yml file


```yml
production:
  adapter: sqlite3
  database: foo.sqlite3
  readwrite: true
```

**Comment by Edouard-chin** [comment]
Actually sorry, adding `readwrite: true` is not a good idea because Rails won't even be able to create the db in legitimate use case. Maybe we can try to add this option to the connection only when we are about to drop the db ?

**Comment by GustavoCaso** [comment]
@Edouard-chin 
> Maybe we can try to add this option to the connection only when we are about to drop the db ?

I'm going to give it a go 😄

**Comment by GustavoCaso** [comment]
@Edouard-chin maybe is my lack of knowledge of the code base but seems that to provide a way to change the option flags when calling `db:drop` to not create the DB when grabbing the current connection `ActiveRecord::Base.connection` will be a more involved change.

I think that checking if the DB exists before trying to acquire the current connection could be the right solution for this issue since it will prevent us from doing unnecessary work since the DB does not exist. 

What do you think? Also, if you think changing the flags is a better option and you want to point to some of the places I should look into it, I'm more than happy to give another go.

**Comment by Edouard-chin** [comment]
I dug on this and although your changes are good (few unimportant nitpicks), I think the clean solution is to go with with the `SQLITE_OPEN_READWRITE` aka `readwrite: true` (at least for sqlite, I haven't dug on postgres but I'm pretty sure it's very similar).

The only reason I thought having `readwrite: true` wasn't possible is because otherwise Rails wouldn't be able to create the database in legitimate use case (i.e. when doing `rails db:create`), but we can bypass this issue by modifying the way the database is created in sqlite (which by the way is really implicit, calling `connections` creates the db ...) https://github.com/rails/rails/blob/722259444ae3a2dbdea7df50b4652e9d4aaaccdb/activerecord/lib/active_record/tasks/sqlite_database_tasks.rb#L21 and just replace it with something like `FileUtils.touch(db_config.database)` (we'd need something better to make sure sub directories are created) . I tried in a dummy app and it works correctly.

So to me the way to go is to always instantiate a connection with `readwrite: true` (with the exception if someone added a `readonly: true` or `flags` in that case we shouldn't add readwrite)
https://github.com/rails/rails/blob/722259444ae3a2dbdea7df50b4652e9d4aaaccdb/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L35

and then fix the issue I mentioned related to creating a database
